## numpy
*numpy is extremely popular python module. It is heavily used in scientific computing.*
**numpy most useful feature is n dimensional array object**
3 main benefits of numpy array over python list are:
1. less memory
2. fast
3. convinient

In [4]:
import numpy as np
a=np.array([1,2,3])
a

array([1, 2, 3])

In [5]:
a[0]

1

In [6]:
a[2]

3

In [13]:
import numpy as np
import time
import sys
l=range(1000)
print(sys.getsizeof(5)*len(l))

array =np.arange(1000)
print(array.size*array.itemsize)


14000
4000


From the above example it is clear that python numpy require less memory as compare to python list

In [14]:
import numpy as np
import time
import sys

SIZE=1000000

l1= range(SIZE)
l2= range(SIZE)

a1=np.arange(SIZE)
a2=np.arange(SIZE)

#python List
start = time.time()
result = [(x+y) for x,y in zip(l1,l2)]
print("python list took:", (time.time()-start)*1000)

#numpy array
start= time.time()
result =a1+a2
print("numpy took:", (time.time()-start)*1000)

python list took: 335.0987434387207
numpy took: 42.127370834350586


python numpy is faster than python list as numpy require 42.127 msec to execute

In [15]:
a1=np.array([1,2,3])
a2=np.array([4,5,6])
a1+a2

array([5, 7, 9])

In [16]:
a1-a2

array([-3, -3, -3])

In [17]:
a1*a2

array([ 4, 10, 18])

In [18]:
a1/a2

array([0.25, 0.4 , 0.5 ])